ykrt/src/location.rs

```diff
@@ -97,88 +29,408 @@ impl State {
     //
-    // We hope that a Location soon reaches PHASE_COMPILED (aka "the happy state") and stays there.
+    //   payload       | IS_COUNTING | IS_PARKED | IS_LOCKED | Notes
+    //   --------------+-----------+---------+---------+-----------------------------------
+    //   <count>       | 1           | 0         | 0         | Start state
+    //                 | 1           | 0         | 1         | Illegal state
+    //                 | 1           | 1         | 0         | Illegal state
+    //                 | 1           | 1         | 1         | Illegal state
+    //   <HotLocation> | 0           | 0         | 0         | Not locked, no-one waiting
+    //   <HotLocation> | 0           | 0         | 1         | Locked, no thread(s) waiting
+    //   <HotLocation> | 0           | 1         | 0         | Not locked, thread(s) waiting
+    //   <HotLocation> | 0           | 1         | 1         | Locked, thread(s) waiting
+    //
+    // where `<count>` is an integer and `<HotLocation>` is a boxed `HotLocation` enum.
+    //
+    // The precise semantics of locking and, in particular, parking are subtle: interested readers
+    // are directed to https://github.com/Amanieu/parking_lot/blob/master/src/raw_mutex.rs#L33 for
+    // a more precise definition.
     state: AtomicUsize,
     phantom: PhantomData<I>,
 }
 
 impl<I> Location<I> {
+    /// Create a new location.
     pub fn new() -> Self {
+        // Locations start in the counting state with a count of 0.
         Self {
-            state: AtomicUsize::new(PHASE_COUNTING),
+            state: AtomicUsize::new(State::<I>::new().x),
             phantom: PhantomData,
         }
     }
 
-    pub(crate) fn load(&self, order: Ordering) -> State {
-        State(self.state.load(order))
+    /// Return this Location's internal state.
+    pub(crate) fn load(&self, order: Ordering) -> State<I> {
+        State {
+            x: self.state.load(order),
+            marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn compare_exchange_weak(
+        &self,
+        current: State<I>,
+        new: State<I>,
+        success: Ordering,
+        failure: Ordering,
+    ) -> Result<State<I>, State<I>> {
+        match self
+            .state
+            .compare_exchange_weak(current.x, new.x, success, failure)
+        {
+            Ok(x) => Ok(State {
+                x,
+                marker: PhantomData,
+            }),
+            Err(x) => Err(State {
+                x,
+                marker: PhantomData,
+            }),
+        }
+    }
+
+    /// Locks this `State` with `Acquire` ordering. It is undefined behaviour to call this function
+    /// if this `State` is in the counting state. If the location moves to the Counting state
+    /// during execution, this will return `Err`.
+    pub(crate) fn lock(&self) -> Result<State<I>, ()> {
+        {
+            let ls = self.load(Ordering::Relaxed);
+            debug_assert!(!ls.is_counting());
```

Comment:
Compared to the rest of this function. Especially the atomic accesses, I think a single check in release mode too is cheap.

Reply:
This is a potentially performance critical part of the code. In the longer run, we'll probably follow parking_lot in splitting this function into an inlinable and non-inlinable component but I can't do that until I have benchmarks, which don't currently run. To be honest, this is more a case of "use `debug_assert` as a form of documentation."

Reply:
Especially since debug assertions are on by default, since we don't (and can't) build the external workspace in release mode.

(This has been bugging me for some time)